Julia1.0 Support

mattBrzezinski opened the pull request:
Summary
---

- Added Julia `1.0`, `1.1`, `1.2` CI jobs, dropped Julia `0.6`
- Dropped Windows CI support (leaving Windows Julia1.2, in case the issue gets resolved) #18 
- Moved from `REQUIRE` to `Project.toml`
- Dropped `build` dir
- Modified unit tests to be independent of each other
- Removed unit tests which didn't really test anything
- Changed whitespace strings to empty strings, replaced whitespace with missing values in `data/Generator_data.dat`, `data/Generator_coord.dat`
- PandaPowers moved from `kW` to `mW` sometime between `v1.4.0 -> v2.1.0`
- Updated to PyCall 1.9.0, resolves: https://github.com/invenia/SyntheticGrids.jl/issues/14

Notes
---
As part of this migration I came across a strange behaviour in the `Basic Grid Creation` unit test, the objects were the same however values were off slightly. After looking around I wasn't able to determine the root cause of this, if anyone has more experience or an idea of where this may have come from let me know.

Original Values:
```
@test mean_shortest_path(adjacency(grid)) ≈ 6.764218612615115	
@test mean_shortest_path(adjacency(grid), distance(buses(grid))) ≈ 133.1803034218734
```

Modified Values:
```
@test mean_shortest_path(adjacency(grid)) ≈ 7.693646165949373
@test mean_shortest_path(adjacency(grid), distance(buses(grid))) ≈ 133.34235584030938
```

After looking into this, the root cause is how Julia1.0 generates random numbers compared to Julia0.6. Here's a sample of the differences between RNG in a similar fashion to this project:

Julia0.6
```
a = srand!(1);
prec = 1.0e-6
rr = 0:prec:(1-prec)

rand(rr)
> 0.745198
rand(rr)
> 0.958146
rand(rr)
> 0.617959
rand(rr)
> 0.944769
rand(rr)
> 0.270773
```

Julia1.0
```
a = Random.seed!(1);
prec = 1.0e-6
rr = 0:prec:(1-prec)

rand(rr)
> 0.484482
rand(rr)
> 0.246119
rand(rr)
> 0.521409
rand(rr)
>0.626485
rand(rr)
> 0.804078
```

Testing:
---

Built the project locally and ran all unit tests successfully.


Quick Links:
---
- [Julia 0.6](https://docs.julialang.org/en/v0.6/)
- [Julia 1.1](https://docs.julialang.org/en/v1.1)

- [PandaPower 1.4.0](https://pandapower.readthedocs.io/en/v1.4.0/elements/load.html)
- [PandaPower 2.1.0](https://pandapower.readthedocs.io/en/v2.1.0/elements/load.html)

- [CSV 0.5 Release Notes](https://discourse.julialang.org/t/ann-new-csv-jl-0-5-release/23848)

mattBrzezinski committed:
Removed build dir

mattBrzezinski committed:
Added deps/build.jl to .gitignore

mattBrzezinski committed:
Cleaned up unit tests to not be dependent on each other

mattBrzezinski committed:
Removed unused filter function

mattBrzezinski committed:
Updated to Pycall 1.9.0; ignoring various build artifacts

mattBrzezinski committed:
Removed generated dummy.json file

mattBrzezinski committed:
Updated source for Julia1.0

mattBrzezinski committed:
Cleaned up data files for Julia1.0

mattBrzezinski committed:
Removed REQUIRE added Project.toml, git ignore artifacts, updated CI runners

mattBrzezinski committed:
Unit tests passing on Julia1.0

mattBrzezinski committed:
Re-structured gitignore

mattBrzezinski committed:
Added version and license to Project.toml

mattBrzezinski committed:
Added case for LLVM v3.5.0

mattBrzezinski committed:
Added Statistics, LinearAlgebra, SparseArrays, Random to SyntheticGrids.jl, removed imports where applicable

mattBrzezinski committed:
Attempting to get Travis CI to pass

eperim commented on Project.toml:
```diff
@@ -0,0 +1,30 @@
+HDF5 = "0.12.0"
+JSON = "0.21.0"
+Missings = "0.4.1"
+PyCall = "1.90.0"
```
Shouldn't we at least leave room for bug fixes in the requirements?

eperim commented on src/io.jl:
```diff
@@ -170,7 +168,7 @@ function get_gen_data(
         "Time from Cold Shutdown to Full Load",
         "Status"
     ]
-    df = CSV.read(datapath, rows_for_type_detect=typemax(Int))
```
IIRC, the `rows_for_type_detect=typemax(Int)` used to be necessary in 0.6 for it to determine types that were general enough. Can you check if it is not choosing types that are too constrained without it?

mattBrzezinski replied:
As of `v0.5` for CSV `rows_for_type_detect` is gone and no longer needed.

> https://discourse.julialang.org/t/ann-new-csv-jl-0-5-release/23848

eperim commented on src/io.jl:
```diff
@@ -223,10 +221,13 @@ function prepare_gen_data(
-            format_gen!(g)
+        plant_dict = [Dict(k => v for (k,v) in zip(keys,g)) for g in p]
+
+        for plant in plant_dict
```
Minor naming issue here, but a `plant_dict` represents a plant is a collection of generators, not of plants (which can either be a single or multiple generators). I'd use `for gen in plant_dict` or similar.

eperim commented on src/io.jl:
```diff
@@ -497,78 +431,55 @@ Currently, grid voltages and line properties are ignored. This function places a
-    const LOAD_VOLT = 110 # kV # This is a crude approximation
-    const GEN_VOLT = 380 # kV # This is a crude approximation
+    LOAD_VOLT = 110 # kV, this is a crude approximation
+    GEN_VOLT = 380 # kV, this is a crude approximation
```
If PandaPower has moved from kW to MW, these constants must be updated accordingly.

eperim commented on src/io.jl:
```diff
@@ -497,78 +431,55 @@ Currently, grid voltages and line properties are ignored. This function places a
                 pgrid,
                 bus = (i-1), # Adopting zero indexing because this is for Python.
-                p_kw = 1000*buses(grid)[i].load, # MW to kW
+                p_mw = buses(grid)[i].load, # MW to kW
```
Probably should remove the comment, as there is no conversion happening anymore.

eperim commented on src/io.jl:
```diff
@@ -497,78 +431,55 @@ Currently, grid voltages and line properties are ignored. This function places a
                     pgrid,
                     bus = (i-1), # Adopting zero indexing because this is for Python.
-                    p_kw = -1000 * gen.cap, # negative by pandapower convention
+                    p_mw = gen.cap * -1,  # Negative for generation
```
`-gen.cap`?

eperim commented on src/io.jl:
```diff
@@ -497,78 +431,55 @@ Currently, grid voltages and line properties are ignored. This function places a
                     name = gen.tech,
-                    max_p_kw = -1000 * gen.minload, # min inverts with max because negative
-                    min_p_kw = -1000 * gen.cap,
+                    max_p_mw = gen.minload * -1,  # Negative for generation
```
it is not just a matter of being negative here. Besides flipping the sign, we are inverting `min` and `max`.

eperim commented on test/runtests.jl:
```diff
@@ -1,136 +1,136 @@
-        )
-        connect!(grid1)
-        create_lines!(grid1)
-        @test grid == grid1
```
This test should not be removed. It tests both that the seeding is working fine and that the custom comparison between grid objects is working as well.

eperim commented on test/runtests.jl:
```diff
@@ -1,136 +1,136 @@
+            @test length(pgrid.trafo) == 31
             pgrid = SyntheticGrids.load_pp_grid(DUMMYPPC)
-            @test length(pgrid[:trafo]) == 36
+            @test length(pgrid.trafo) == 31
```
This has to be looked into, as the new test is finding 5 fewer transformers from the data.

mattBrzezinski replied:
Posting for posterity here, will also update the description and post sample code there.

Yesterday going through we were able to find that generating random numbers has differed between Julia0.6 and Julia1.0. This results in the adjacency matrices being different in versions and then changing the amount of transformers.

eperim commented:
> As part of this migration I came across a strange behaviour in the Basic Grid Creation unit test, the objects were the same however values were off slightly. After looking around I wasn't able to determine the root cause of this, if anyone has more experience or an idea of where this may have come from let me know.

I haven't looked into it, but maybe it has to do with the random number generator working differently in 1.0 from 0.6? This is just a blind guess, no idea if actually there any difference there.

mattBrzezinski committed:
Allow deps to have minor version bumps

mattBrzezinski committed:
Added unit test checking seeding and obj comparison, minor variable renames, LOAD and GEN voltages to mV

mattBrzezinski committed:
Removed whitespace, modified baseline_grid to use variables

mattBrzezinski committed:
Added Test, Documenters to CI jobs

Added Conda.BINDIR to ENV[PATH] so DLLs map properly on Windows

CondaForge + NumPy1.17.0, Write build.log out

Empty str instead of blank space in deps/build.jl

Dropped appveyor, added Travis CI windows

Downgraded HDF5 to v0.10.X, Blosc (a dep) is having issues on Windows building

Explicitly set BinaryProvider one minor version than current (0.5.X)

mattBrzezinski committed:
Travis CI modified to ignore Windows builds except for Julia 1.2 (latest version) and allow failures

nickrobinson251 commented on .travis.yml:
```diff
@@ -1,45 +1,50 @@
+  - if [[ -a .git/shallow ]]; then git fetch --unshallow; fi
+  - |
+    julia -e '
+      VERSION >= v"0.7.0-DEV.3656" && using Pkg
```
According to the Project.toml, we don't support `VERSION < v"1"`, so all the `VERSION >= v"0.7.0` checks can be removed

nickrobinson251 commented on src/bus.jl:
```diff
@@ -12,15 +12,15 @@ struct LoadBus <: Bus
         connected_to = Set(),
         connections = Set()
-    )
+    ) where T <: Real
```
why aren't `LoadBus` and `GenBus` parameterised by the coord type anymore? I mean, it seems like a fine change to me, but I don't use the package so I just want to flag it...

eperim replied:
They never really were, that was just old 0.6 syntax. In the old code it was the constructors that had the type declaration, which now works with `where ...` instead. The objects have never been parametric.

nickrobinson251 replied:
Thanks for the clarification! I misread this as changing the types parameterisation, as you noticed.

nickrobinson251 commented on src/clustering.jl:
```diff
@@ -25,7 +25,7 @@ function cluster_loads!(grid::Grid, nload)
     id2ind = Dict{Int, Int}()
     for id in all_ids
-        id2ind[id] = findfirst(all_ids, id)
+        id2ind[id] = findfirst(x->x==id, all_ids)
```
```suggestion
        id2ind[id] = findfirst(==(id), all_ids)
```

nickrobinson251 commented on src/clustering.jl:
```diff
@@ -34,21 +34,21 @@ function cluster_loads!(grid::Grid, nload)
                 break
             end
-            i, j = ind2sub(size(sub_dist), m)
+            i, j = Tuple(CartesianIndices(size(sub_dist))[m])
```
What is this line doing? I don't understand what it was doing before (I thought `ind2sub` was a `Dict`, but `Dict`s are not callable like `ind2sub(x, y)`), but my worry is this has different behaviour...

eperim replied:
It takes a linear index and turns into the 2D index of a matrix. e.g.: If one has a 2x4 matrix, it maps 5 to (2, 1).

nickrobinson251 replied:
well, it does select the index, but it iterates column-wise (not row-wise) -- just checking this is the behaviour you want
```julia
julia> X = rand(2, 4);

julia> CartesianIndices(size(X))
2×4 CartesianIndices{2,Tuple{Base.OneTo{Int64},Base.OneTo{Int64}}}:
 CartesianIndex(1, 1)  CartesianIndex(1, 2)  CartesianIndex(1, 3)  CartesianIndex(1, 4)
 CartesianIndex(2, 1)  CartesianIndex(2, 2)  CartesianIndex(2, 3)  CartesianIndex(2, 4)

julia> CartesianIndices(size(X))[5]
CartesianIndex(1, 3)
```

eperim replied:
Botched example of mine. Good thing I was not the one who implemented `CartesianIndices` =P

omus replied:
Can't this be simplified to:
```julia
i, j = Tuple(m)
```
It used to be that `indmin` returned a linear index but now both `findmin` and `argmin` return `CartesianIndex`

nickrobinson251 commented on src/distances.jl:
```diff
@@ -10,13 +12,13 @@ end
 
 """
-    haversine{T<:Real}(c1::Tuple{T,T}, c2::Tuple{T,T})
+    haversine(c1::Tuple{T,T}, c2::Tuple{T,T}) where T <: Real
```
👍

eperim replied:
This is the same case as the one you raised in the `LoadBus`/`GenBus` case. It is the same syntax change.

nickrobinson251 replied:
it's a good change! I misread the case with the constructor above, as you duly noticed 😆

nickrobinson251 commented on src/io.jl:
```diff
@@ -135,11 +136,8 @@ function get_plant_data(coordpath = GENCOORDPATH)
-            Symbol("Grid Voltage 3 (kV)")
-        ]
+
+        for s in [Symbol("Grid Voltage (kV)"), Symbol("Grid Voltage 2 (kV)"), Symbol("Grid Voltage 3 (kV)")]
```
If these are all the column names

```suggestion
        for s in names(df)
```
otherwise
```suggestion
        for s in keys(column_types)
```

eperim replied:
I believe `df` has other columns, so we can't simply call `names` or `keys`.

nickrobinson251 replied:
we can use `column_types` because we know that has these names 
https://github.com/invenia/SyntheticGrids.jl/pull/17/files#diff-b5ddd5a415e261ce7d1b14e8bb8f9dfaR127

nickrobinson251 commented on src/io.jl:
```diff
@@ -223,10 +221,13 @@ function prepare_gen_data(
-        push!(fplants, [Dict(k => v for (k,v) in zip(keys,g)) for g in p])
-        for g in fplants[end]
-            format_gen!(g)
+        plant_dict = [Dict(k => v for (k,v) in zip(keys,g)) for g in p]
```
```suggestion
        plant_dict = (Dict(zip(_keys, g)) for g in p)
```

nickrobinson251 commented on src/io.jl:
```diff
@@ -223,10 +221,13 @@ function prepare_gen_data(
+            format_gen!(gen)
         end
+
+        push!(fplants, plant_dict)
```
```suggestion
        push!(fplants, collect(plant_dict))
```

eperim replied:
Each `fplant` is one Dict. We want a vector of dicts, not one large dict.

nickrobinson251 replied:
yeah, sorry, this is only an acceptable change if we change it to a Generator above (i.e. do both or do neither) https://github.com/invenia/SyntheticGrids.jl/pull/17#discussion_r313322688

nickrobinson251 replied:
we know need this for the behaviour to be correct :)

nickrobinson251 commented on src/io.jl:
```diff
@@ -223,10 +221,13 @@ function prepare_gen_data(
     keys = ["coords", "volt", "tech", "cap", "pfactor", "minload", "scap",
```
`keys` is a function, so best to avoid that name
```suggestion
     _keys = ["coords", "volt", "tech", "cap", "pfactor", "minload", "scap",
```

nickrobinson251 reviewed (commented):
Looks good - couple minor things, mostly questions stemming from me not knowing the package. If @eperim is fine with it, so am i :)

eperim commented:
> Looks good - couple minor things, mostly questions stemming from me not knowing the package. If @eperim is fine with it, so am i :)

It is all good to me, just really waiting for the CI issues to be solved.

mattBrzezinski committed:
Addressed CR feedback, minor var renaming, removal of 0.7 checks in Travis script

Removed AppVeyor badge

omus commented on Project.toml:
```diff
@@ -0,0 +1,32 @@
+Random = "9a3f8284-a2c9-5f02-9a11-845980a1fd5c"
+SparseArrays = "2f01184e-e22b-5df5-ae63-d93ebab69eaf"
+Statistics = "10745b16-79ce-11e8-11f9-7d13ad32a3b2"
+Test = "8dfed614-e22c-5e08-85e1-65c5234f0b40"
```
I suspect this is only a test dependency

omus commented on deps/build.jl:
```diff
@@ -1,34 +1,40 @@
 using Conda
 
-ENV["PYTHON"]=" "  # Configure PyCall.jl to use Conda.jl's Python
+ENV["PYTHON"]=""  # Configure PyCall.jl to use Conda.jl's Python
```
```suggestion
ENV["PYTHON"] = ""  # Configure PyCall.jl to use Conda.jl's Python
```

omus commented on src/SyntheticGrids.jl:
```diff
@@ -4,18 +4,24 @@ module SyntheticGrids
+using SparseArrays
+using Random
+using Test
+using Documenter
```
Where are `Documenter` and `Test` used?

omus commented on src/bus.jl:
```diff
@@ -36,14 +36,14 @@ function LoadBus(coords::LatLon, load, volt, pop, connected_to = Set(), connecti
 
-function LoadBus{T <: Real}(
+function LoadBus(
         coords::Tuple{T,T},
```
If `T` is not used in the function body you should be able to write this as:
```suggestion
        coords::Tuple{T,T} where T <: Real,
```

omus commented on src/clustering.jl:
```diff
@@ -34,21 +34,21 @@ function cluster_loads!(grid::Grid, nload)
         i = j = -1
         while true
-            m = indmin(sub_dist)
+            m = findmin(sub_dist)[2]  # [2] is the index of the smallest item
```
`indmin` was replaced with `argmin`:
```suggestion
            m = argmin(sub_dist)
```

omus commented on src/clustering.jl:
```diff
@@ -96,8 +97,9 @@ Create 'nboth' substations with both load and generation, chosen randomly.
     gens = [bus for bus in buses(grid) if isa(bus, GenBus)]
     nb = 0
-    srand(grid.seed + 66) # Just so we don't keep returning to the same point.
+    Random.seed!(grid.seed + 66) # Just so we don't keep returning to the same point.
```
Instead of doing this I would recommend doing:
```julia
using Random: seed!
```
As this lets developers know what functions you are using from `Random`

omus commented on src/distances.jl:
```diff
@@ -1,3 +1,5 @@
+import Statistics: mean
```
```suggestion
using Statistics: mean
```
Unless you are adding methods to `mean`

omus commented on src/distances.jl:
```diff
@@ -36,7 +38,7 @@ function haversine(b1::Bus, b2::Bus)
 end
 
-function mean{T <: Real}(v::Vector{LatLon{T}})
+function mean(v::Vector{LatLon{T}}) where T <: Real
```
```suggestion
function mean(v::Vector{LatLon{<:Real}})
```

omus commented on src/distances.jl:
```diff
@@ -51,13 +53,13 @@ REFERENCE: Birchfield, Adam B., et al. "Grid Structural Characteristics as
+function subs_dist(
     inds1::Vector{Int},
     inds2::Vector{Int},
     b_dist::Matrix{T},
```
```suggestion
    b_dist::Matrix{<:Real},
```

omus commented on src/distances.jl:
```diff
@@ -122,7 +124,7 @@ function update_sub_dist!(
 end
 
-function distance{T<:Bus}(buses::Vector{T})
+function distance(buses::Vector{T}) where T <: Bus
```
```suggestion
function distance(buses::Vector{<:Bus})
```

omus commented on src/distances.jl:
```diff
@@ -132,7 +134,7 @@ function distance{T<:Bus}(buses::Vector{T})
 end
 
-function update_distance!{T<:Bus}(dist_mat::Matrix{Float64}, buses::Vector{T}, i::Int)
+function update_distance!(dist_mat::Matrix{Float64}, buses::Vector{T}, i::Int) where T <: Bus
```
```suggestion
function update_distance!(dist_mat::Matrix{Float64}, buses::Vector{<:Bus}, i::Int)
```

omus commented on src/generator.jl:
```diff
@@ -13,7 +13,7 @@
 
-function Generator{T <: Real}(
+function Generator(
         coords::Tuple{T,T},
```
```suggestion
        coords::Tuple{T,T} where T <: Real,
```

omus commented on src/io.jl:
```diff
@@ -123,8 +123,9 @@ function place_loads_from_zips!(
-function get_plant_data(coordpath = GENCOORDPATH)
-    df = CSV.read(coordpath, rows_for_type_detect=typemax(Int))
+function get_plant_data(coordpath=GENCOORDPATH)
+    column_types = Dict("Grid Voltage (kV)" => Float64, "Grid Voltage 2 (kV)" => Float64, "Grid Voltage 3 (kV)" => Float64)
```
Line is a bit long:
```suggestion
    column_types = Dict(
        "Grid Voltage (kV)" => Float64,
        "Grid Voltage 2 (kV)" => Float64,
        "Grid Voltage 3 (kV)" => Float64,
    )
```

omus commented on src/substation.jl:
```diff
@@ -13,7 +13,7 @@ mutable struct Substation <: Bus
-function Substation{T<:Real}(
+function Substation(
     id,
     coords::Tuple{T,T},
```
```suggestion
    coords::Tuple{T,T} where T <: Real,
```

omus commented on src/substation.jl:
```diff
@@ -39,15 +39,15 @@ function Substation(coords::LatLon, volts, load, gen, pop, con = Set(), group =
 
-function Substation{T<:Real}(
+function Substation(
     coords::Tuple{T,T},
```
```suggestion
    coords::Tuple{T,T} where T <: Real,
```

omus commented on test/runtests.jl:
```diff
@@ -1,136 +1,143 @@
+    SEED = 666
+    LATLIM = (38, 40)
+    LONGLIM = (-89, -88)
+    GENPATH = joinpath(dirname(@__FILE__), "..", "data", "GenData.json")
```
You can replace `dirname(@__FILE__)` with `@__DIR__`:
```suggestion
    GENPATH = joinpath(@__DIR__, "..", "data", "GenData.json")
```

omus commented on test/runtests.jl:
```diff
@@ -1,136 +1,143 @@
             pgrid = SyntheticGrids.load_pp_grid(DUMMYPPC)
-            @test length(pgrid[:trafo]) == 36
+            @test length(pgrid.trafo) == 31
+            rm(DUMMYPPC)
```
If this testset were to fail and this file was not removed what would happen?

mattBrzezinski replied:
Nothing really, you'd just have the artifact from the test left over. We just write to this file so that we can subsequently load from it to ensure that PandaPower is saving/opening properly.

If the testset fails, every other test would overwrite this artifact in line 118.

omus commented:
I've disabled the Appveyor webhook on this repo. I closed/re-opened the PR to see if it would restart the CI.

omus commented on src/clustering.jl:
```diff
@@ -79,7 +82,7 @@ until generation threshold 'MW' is reached.
     dists = [haversine(sub, g) for g in gens]
     while sub.generation < MW
-        m = indmin(dists)
+        m = findmin(dists)[2]  # [2] is the index of the smallest item
```
```suggestion
        m = argmin(dists)
```

omus commented on src/clustering.jl:
```diff
@@ -140,15 +144,15 @@ function cluster_gens!(grid::Grid, ngen)
     dists = distance(g_subs)
     while length(g_subs) > ngen
-        m = indmin(dists)
+        m = findmin(dists)[2]  # [2] is the index of the smallest item
```
```suggestion
        m = argmin(dists)
```

omus commented on src/clustering.jl:
```diff
@@ -140,15 +144,15 @@ function cluster_gens!(grid::Grid, ngen)
+        m = findmin(dists)[2]  # [2] is the index of the smallest item
         n = length(g_subs)
-        ii, jj = ind2sub(size(dists), m)
+        ii, jj = Tuple(CartesianIndices(size(dists))[m])
```
```suggestion
        ii, jj = Tuple(m)
```

omus commented on src/grid.jl:
```diff
@@ -237,9 +233,9 @@ function twst!(grid::Grid, k)
+    permut = Vector{Int}(undef, n)
     unorm_prob = [haversine(bus.coords, mean(node_locations))^(-k) for bus in buses(grid)]
-    srand(grid.seed + 15)
+    Random.seed!(grid.seed + 15)
```
```suggestion
   seed!(grid.seed + 15)
```

omus commented on src/grid.jl:
```diff
@@ -314,7 +310,7 @@ function reinforce!(grid::Grid, m, a, b, g, t, N)
         inds = collect(1:n)
     end
-    srand(grid.seed + 479) # Just so we don't keep returning to the same point.
+    Random.seed!(grid.seed + 479) # Just so we don't keep returning to the same point.
```
```suggestion
   seed!(grid.seed + 479) # Just so we don't keep returning to the same point.
```

omus commented on src/grid.jl:
```diff
@@ -416,7 +412,7 @@ This function will only work after the grid has been connected through the
 """
 function create_lines!(grid::Grid; impedfunc = linear_imped, capfunc = volt_cap)
-    srand(grid.seed + 16) # Just so we don't keep returning to the same point.
+    Random.seed!(grid.seed + 16) # Just so we don't keep returning to the same point.
```
```suggestion
   seed!(grid.seed + 16) # Just so we don't keep returning to the same point.
```

omus commented on src/io.jl:
```diff
@@ -60,7 +60,7 @@ function place_loads_from_zips!(
     ]
 
-    srand(grid.seed + 23) # Just so we don't keep returning to the same point.
+    Random.seed!(grid.seed + 23) # Just so we don't keep returning to the same point.
```
```suggestion
   seed!(grid.seed + 23) # Just so we don't keep returning to the same point.
```

omus commented on src/io.jl:
```diff
@@ -107,7 +107,7 @@ function place_loads_from_zips!(
     fzips = [z for z in allzips if geolim((z[3], z[4]))]
 
-    srand(grid.seed + 23) # Just so we don't keep returning to the same point.
+    Random.seed!(grid.seed + 23) # Just so we don't keep returning to the same point.
```
```suggestion
   seed!(grid.seed + 23) # Just so we don't keep returning to the same point.
```

omus commented on src/SyntheticGrids.jl:
```diff
@@ -73,7 +79,7 @@ load_grid
 import Base.show
 import Base.hash
-import Base.mean
+import Statistics.mean
```
I would just do this in `src/SyntheticGrids.jl` and remove the `using` statements from the other files.
```suggestion
using Statistics: Statistics, mean
```

omus commented on src/clustering.jl:
```diff
@@ -1,7 +1,9 @@
+using Random: seed!
```
Just put this in `src/SyntheticGrids.jl`

omus commented on src/distances.jl:
```diff
@@ -1,3 +1,5 @@
+using Statistics: Statistics, mean
```
Can be removed.
```suggestion
```

omus commented on src/grid.jl:
```diff
@@ -1,3 +1,5 @@
+using Statistics: Statistics, mean
```
Can be removed.
```suggestion

```

omus commented on src/SyntheticGrids.jl:
```diff
@@ -73,8 +79,10 @@ load_grid
-import Base.mean
 import Base.merge!
+using Random: seed!
+using Statistics: Statistics, mean
```
Should be combined with the above `using` calls. Ideally you only call `using Statistics` once

omus commented on Project.toml:
```diff
@@ -0,0 +1,37 @@
+Geodesy = "^0.5.0"
+HDF5 = "^0.12.0"
+JSON = "^0.21.0"
+Missings = "^0.4.0"
```
Missing support is included in Base as of Julia 1.0

omus replied:
You should also remove this from `src/SyntheticGrids.jl` and `[deps]` in the `Project.toml`

omus commented on Project.toml:
```diff
@@ -0,0 +1,37 @@
+AutoHashEquals = "15f4f7f2-30c1-5605-9d31-71845cf9641f"
+CSV = "336ed68f-0bac-5ca0-87d4-7b16caf5d00b"
+Conda = "8f4d0f93-b110-5947-807f-2305c1781a2d"
+Documenter = "e30172f5-a6a5-5a46-863b-614d45cd2de4"
```
Not used in `src` so can be removed

omus commented on Project.toml:
```diff
@@ -0,0 +1,37 @@
+Conda = "8f4d0f93-b110-5947-807f-2305c1781a2d"
+Documenter = "e30172f5-a6a5-5a46-863b-614d45cd2de4"
+Geodesy = "0ef565a4-170c-5f04-8de2-149903a85f3d"
+HDF5 = "f67ccb44-e63f-5c2f-98bd-6dc0ccc4ba2f"
```
Unused and can be removed

omus commented on src/SyntheticGrids.jl:
```diff
@@ -4,18 +4,23 @@ module SyntheticGrids
 # at load time ? :/
 ENV["NUMBA_DISABLE_JIT"] = 1
 
+using Statistics
```
Can this be removed?

mattBrzezinski commented:
@JuliaRegistrator register

JuliaRegistrator commented:
Error while trying to register: **Register Failed**
@mattBrzezinski, it looks like you are not a publicly listed member/owner in the parent organization (invenia).
If you are a member/owner, you will need to change your membership to public. See [GitHub Help](https://help.github.com/en/articles/publicizing-or-hiding-organization-membership)

JuliaRegistrator commented:
Pull request comments will not trigger Registrator as it is disabled. Please trying using a commit or issue comment.

mattBrzezinski commented:
@JuliaRegistrator register

JuliaRegistrator commented:
Pull request comments will not trigger Registrator as it is disabled. Please trying using a commit or issue comment.
